File: ui.py
"""Loopback-only, single-user browser UI. No additional server dependencies."""

import json
import re
import secrets
import shutil
import tempfile
import threading
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlsplit

from app import EXTENSIONS, SearchablePDF
from engine import OCR, overlay, pages
from native import ROOT

MAX_UPLOAD = 25 * 1024 * 1024
MAX_PAGES = 20


class UIServer(ThreadingHTTPServer):
    daemon_threads = True

    def __init__(self, address, model, workspace):
        self.ocr = OCR(model)
        self.workspace = Path(workspace)
        self.token = secrets.token_hex(32)
        self.work_lock = threading.Lock()
        super().__init__(address, Handler)

    def prune(self):
        jobs = sorted(self.workspace.iterdir(), key=lambda p: p.stat().st_mtime)
        for index, job in enumerate(jobs):
            if time.time() - job.stat().st_mtime > 3600 or index < len(jobs) - 9:
                shutil.rmtree(job, ignore_errors=True)


class Handler(BaseHTTPRequestHandler):
    server_version = "MojoOCR/0.1.0"

    def setup(self):
        super().setup()
        self.connection.settimeout(60)

    def log_message(self, fmt, *args):
        # Do not log document names, result tokens, or text.
        pass

    def send_headers(self, status, content_type, length, download=None):
        self.send_response(status)
        self.send_header("Content-Type", content_type)
        self.send_header("Content-Length", str(length))
        self.send_header("Cache-Control", "no-store")
        self.send_header("X-Content-Type-Options", "nosniff")
        self.send_header("Referrer-Policy", "no-referrer")
        self.send_header(
            "Content-Security-Policy",
            "default-src 'self'; script-src 'self'; style-src 'self'; img-src 'self' blob:; connect-src 'self'; object-src 'none'; base-uri 'none'; frame-ancestors 'none'",
        )
        if download:
            self.send_header("Content-Disposition", f'attachment; filename="{download}"')
        self.end_headers()

    def json(self, status, value):
        body = json.dumps(value).encode()
        self.send_headers(status, "application/json; charset=utf-8", len(body))
        self.wfile.write(body)

    def file(self, path, content_type, download=None):
        try:
            with path.open("rb") as stream:
                self.send_headers(200, content_type, path.stat().st_size, download)
                shutil.copyfileobj(stream, self.wfile)
        except FileNotFoundError:
            self.json(404, {"error": "Result expired or cleared. Run OCR again."})

    def trusted(self, mutation=False):
        port = self.server.server_port
        hosts = {f"localhost:{port}", f"127.0.0.1:{port}"}
        if self.headers_in.get("Host") not in hosts:
            return False
        origin = self.headers_in.get("Origin")
        if origin and origin not in {"http://" + host for host in hosts}:
            return False
        if self.headers_in.get("Sec-Fetch-Site") == "cross-site":
            return False
        return not mutation or secrets.compare_digest(
            self.headers_in.get("X-OCR-Token", ""), self.server.token
        )

    def do_GET(self):
        self.headers_in = self.headers
        if not self.trusted():
            self.json(403, {"error": "Only same-origin localhost requests are allowed."})
            return
        path = urlsplit(self.path).path
        static = {
            "/": ("ui.html", "text/html; charset=utf-8"),
            "/ui.js": ("ui.js", "text/javascript; charset=utf-8"),
            "/ui.css": ("ui.css", "text/css; charset=utf-8"),
            "/example.png": ("examples/note.png", "image/png"),
            "/assets/logo.svg": ("assets/logo.svg", "image/svg+xml"),
        }
        if path == "/favicon.ico":
            self.send_headers(204, "image/x-icon", 0)
        elif path in static:
            name, mime = static[path]
            self.file(ROOT / name, mime)
        elif path == "/api/status":
            self.json(
                200,
                {
                    "ready": True,
                    "token": self.server.token,
                    "model": Path(self.server.ocr.model_info["path"]).name,
                    "max_upload_mb": MAX_UPLOAD // (1024 * 1024),
                    "max_pages": MAX_PAGES,
                },
            )
        else:
            match = re.fullmatch(
                r"/files/([a-f0-9]{32})/(page-\d+(?:-boxes)?\.jpg|result\.json|searchable\.pdf)",
                path,
            )
            if not match:
                self.json(404, {"error": "Not found"})
                return
            job, name = match.groups()
            target = self.server.workspace / job / name
            if target.exists() and time.time() - target.parent.stat().st_mtime > 3600:
                self.json(410, {"error": "Result expired. Run OCR again."})
                return
            mime = (
                "image/jpeg"
                if name.endswith(".jpg")
                else "application/pdf"
                if name.endswith(".pdf")
                else "application/json"
            )
            self.file(target, mime, None if name.endswith(".jpg") else name)

    def do_POST(self):
        self.headers_in = self.headers
        self.close_connection = True
        if not self.trusted(mutation=True):
            self.json(403, {"error": "Invalid local session. Reload the page."})
            return
        url = urlsplit(self.path)
        if url.path not in {"/api/ocr", "/api/clear"}:
            self.json(404, {"error": "Not found"})
            return
        if not self.server.work_lock.acquire(blocking=False):
            self.json(409, {"error": "OCR is already running. Please wait and try again."})
            return
        job = None
        try:
            if url.path == "/api/clear":
                for item in self.server.workspace.iterdir():
                    shutil.rmtree(item, ignore_errors=True)
                self.json(200, {"cleared": True})
                return
            length = int(self.headers_in.get("Content-Length", "0"))
            if not 0 < length <= MAX_UPLOAD:
                self.json(413, {"error": "Choose a non-empty file smaller than 25 MiB."})
                return
            query = parse_qs(url.query)

            def get(key, default=""):
                return query.get(key, [default])[0]

            name = get("name").replace("\\", "/").split("/")[-1][:200]
            suffix = Path(name).suffix.lower()
            if suffix not in EXTENSIONS:
                raise ValueError("Choose a PNG, JPG, WebP, BMP, TIFF, or PDF file.")
            dpi, rotate = int(get("dpi", "150")), int(get("rotate", "0"))
            if dpi not in {150, 200, 300} or rotate not in {0, 90, 180, 270}:
                raise ValueError("Invalid image settings")
            self.server.prune()
            key = secrets.token_hex(16)
            job = self.server.workspace / key
            job.mkdir()
            source = job / ("input" + suffix)
            with source.open("wb") as stream:
                remaining = length
                while remaining:
                    chunk = self.rfile.read(min(remaining, 1024 * 1024))
                    if not chunk:
                        raise ValueError("Upload interrupted. Please retry.")
                    stream.write(chunk)
                    remaining -= len(chunk)
            start = time.perf_counter()
            pdf = SearchablePDF(job / "searchable.pdf", dpi)
            results = []
            for number, image in pages(source, dpi=dpi, max_pages=MAX_PAGES):
                result, processed, _ = self.server.ocr.image(
                    image,
                    rotate=rotate,
                    deskew=get("deskew") == "1",
                    adaptive=get("adaptive") == "1",
                    whitelist="0123456789.,-$+%()/" if get("mode") == "numbers" else None,
                )
                pdf.add(processed, result)
                marked = overlay(processed, result)
                for picture, tail in [(processed, ""), (marked, "-boxes")]:
                    picture.thumbnail((1400, 1400))
                    picture.save(job / f"page-{number}{tail}.jpg", quality=88)
                result.update(
                    page=number,
                    preview=f"/files/{key}/page-{number}.jpg",
                    overlay=f"/files/{key}/page-{number}-boxes.jpg",
                )
                results.append(result)
            pdf.close()
            source.unlink()  # Keep derivatives only until cleared/server exit.
            data = {
                "name": name,
                "pages": results,
                "seconds": time.perf_counter() - start,
                "pdf": f"/files/{key}/searchable.pdf",
                "json": f"/files/{key}/result.json",
            }
            (job / "result.json").write_text(json.dumps(data, indent=2))
            self.json(200, data)
        except (BrokenPipeError, ConnectionResetError):
            if job:
                shutil.rmtree(job, ignore_errors=True)
        except Exception as exc:
            if job:
                shutil.rmtree(job, ignore_errors=True)
            self.json(400, {"error": str(exc) or "Unable to read this file."})
        finally:
            self.server.work_lock.release()


def serve(args):
    if not 1 <= args.port <= 65535:
        raise ValueError("--port must be between 1 and 65535")
    with tempfile.TemporaryDirectory(prefix="mojo-ocr-") as workspace:
        server = UIServer(("127.0.0.1", args.port), args.model, workspace)
        print(
            f"Mojo OCR → http://localhost:{server.server_port}\nLocal only. Ctrl+C stops the server and deletes temporary results.",
            flush=True,
        )
        try:
            server.serve_forever()
        except KeyboardInterrupt:
            pass
        finally:
            server.shutdown()
            server.server_close()
            # Wait for an active upload/recognition before removing its workspace.
            with server.work_lock:
                pass

File: ui.js
"use strict";
const $ = (id) => document.getElementById(id);
let token = "",
  selected = null,
  result = null,
  edits = [],
  page = 0,
  busy = false;

function message(text, error = false) {
  if (error) {
    $("status").textContent = "Something needs your attention.";
    $("error").textContent = text;
    $("error").hidden = false;
  } else {
    $("status").textContent = text;
    $("error").hidden = true;
  }
}
function controls() {
  $("extract").disabled = busy || !token || !selected;
  $("clear").disabled = busy || (!result && !selected);
  for (const id of [
    "file",
    "example",
    "mode",
    "rotate",
    "dpi",
    "deskew",
    "adaptive",
  ])
    $(id).disabled = busy;
  $("status").classList.toggle("busy", busy);
  $("extract-label").textContent = busy
    ? "Reading your document…"
    : result
      ? "Extract again"
      : "Extract text";
  $("drop").setAttribute("aria-disabled", String(busy));
}
function select(file) {
  if (busy || !file) return;
  if (!/\.(png|jpe?g|webp|bmp|tiff?|pdf)$/i.test(file.name)) {
    message("Please choose an image or a PDF file.", true);
    return;
  }
  if (!file.size || file.size > 25 * 1024 * 1024) {
    message("Choose a non-empty file smaller than 25 MiB.", true);
    return;
  }
  if (
    result &&
    edits.some((text, i) => text !== result.pages[i].text) &&
    !confirm("Replacing this document discards your text edits. Continue?")
  )
    return;
  selected = file;
  result = null;
  edits = [];
  page = 0;
  $("result").hidden = true;
  $("empty").hidden = false;
  $("result-title").textContent = "Your text starts here";
  $("file-name").textContent = file.name;
  $("file-detail").textContent =
    `${(file.size / 1024).toLocaleString(undefined, { maximumFractionDigits: 0 })} KB · click to replace`;
  message("Ready when you are.");
  controls();
}
$("file").addEventListener("change", () => select($("file").files[0]));
$("drop").addEventListener("keydown", (event) => {
  if (!busy && ["Enter", " "].includes(event.key)) {
    event.preventDefault();
    $("file").click();
  }
});
for (const type of ["dragenter", "dragover"])
  $("drop").addEventListener(type, (event) => {
    event.preventDefault();
    if (!busy) $("drop").classList.add("dragging");
  });
for (const type of ["dragleave", "drop"])
  $("drop").addEventListener(type, (event) => {
    event.preventDefault();
    $("drop").classList.remove("dragging");
  });
$("drop").addEventListener("drop", (event) =>
  select(event.dataTransfer.files[0]),
);
// Avoid the browser navigating away when a file misses the drop target.
window.addEventListener("dragover", (event) => event.preventDefault());
window.addEventListener("drop", (event) => event.preventDefault());
window.addEventListener("paste", (event) => {
  if (busy || ["TEXTAREA", "INPUT"].includes(event.target.tagName)) return;
  const item = [...(event.clipboardData?.items || [])].find((item) =>
    item.type.startsWith("image/"),
  );
  if (item) {
    event.preventDefault();
    const blob = item.getAsFile();
    select(new File([blob], "pasted-image.png", { type: blob.type }));
  }
});
$("example").addEventListener("click", async () => {
  try {
    const response = await fetch("/example.png");
    if (!response.ok) throw Error("Example unavailable.");
    select(
      new File([await response.blob()], "example-note.png", {
        type: "image/png",
      }),
    );
  } catch (error) {
    message(error.message, true);
  }
});
function renderPage() {
  if (!result) return;
  const data = result.pages[page];
  $("text").value = edits[page];
  const image = $("boxes").checked ? data.overlay : data.preview;
  $("preview").src = image;
  $("preview-link").href = image;
  $("legend").hidden = !$("boxes").checked;
  $("metrics").textContent =
    `${data.uncertain_characters} low-confidence glyphs · ${(data.seconds * 1000).toFixed(0)} ms`;
  updateEdited();
}
function updateEdited() {
  $("character-count").textContent =
    `${$("text").value.length.toLocaleString()} characters`;
  $("edited").hidden = !result || edits[page] === result.pages[page].text;
}
$("page").addEventListener("change", () => {
  page = Number($("page").value);
  renderPage();
});
$("boxes").addEventListener("change", renderPage);
$("text").addEventListener("input", () => {
  edits[page] = $("text").value;
  updateEdited();
});
$("extract").addEventListener("click", async () => {
  if (busy || !selected) return;
  if (
    result &&
    edits.some((text, i) => text !== result.pages[i].text) &&
    !confirm("Extracting again replaces your text edits. Continue?")
  )
    return;
  busy = true;
  controls();
  message("Reading locally. Larger PDFs may take a little longer.");
  try {
    const params = new URLSearchParams({
      name: selected.name,
      mode: $("mode").value,
      dpi: $("dpi").value,
      rotate: $("rotate").value,
      deskew: $("deskew").checked ? "1" : "0",
      adaptive: $("adaptive").checked ? "1" : "0",
    });
    const response = await fetch("/api/ocr?" + params, {
      method: "POST",
      headers: {
        "Content-Type": "application/octet-stream",
        "X-OCR-Token": token,
      },
      body: selected,
    });
    const data = await response.json();
    if (!response.ok)
      throw Error(data.error || "OCR failed. Please try again.");
    if (!data.pages?.length) throw Error("No pages found in this document.");
    result = data;
    edits = data.pages.map((p) => p.text);
    page = 0;
    $("result-title").textContent = data.name;
    $("empty").hidden = true;
    $("result").hidden = false;
    $("page").replaceChildren(
      ...data.pages.map(
        (p, i) => new Option(`${i + 1} / ${data.pages.length}`, i),
      ),
    );
    $("download-pdf").href = data.pdf;
    $("download-json").href = data.json;
    renderPage();
    message(
      `Done. ${data.pages.length} page${data.pages.length === 1 ? "" : "s"} in ${data.seconds.toFixed(1)}s. Review before using.`,
    );
  } catch (error) {
    message(
      error.message || "Unable to connect. Is the local server still running?",
      true,
    );
  } finally {
    busy = false;
    controls();
  }
});
$("clear").addEventListener("click", async () => {
  if (
    result &&
    !confirm(
      "Clear this workspace and all temporary server results? Download anything you want to keep first.",
    )
  )
    return;
  busy = true;
  controls();
  try {
    const response = await fetch("/api/clear", {
      method: "POST",
      headers: { "X-OCR-Token": token },
    });
    if (!response.ok) throw Error("Could not clear results. Please try again.");
    result = null;
    selected = null;
    edits = [];
    page = 0;
    $("file").value = "";
    $("text").value = "";
    $("preview").removeAttribute("src");
    $("file-name").textContent = "Drop a file here";
    $("file-detail").textContent = "or click to browse";
    $("result-title").textContent = "Your text starts here";
    $("result").hidden = true;
    $("empty").hidden = false;
    message("Workspace cleared.");
  } catch (error) {
    message(error.message, true);
  } finally {
    busy = false;
    controls();
  }
});
$("copy").addEventListener("click", async () => {
  try {
    await navigator.clipboard.writeText($("text").value);
    $("copy").textContent = "Copied!";
    setTimeout(() => {
      $("copy").textContent = "Copy text";
    }, 1800);
  } catch {
    $("text").focus();
    $("text").select();
    message("Text selected. Press Ctrl+C to copy.");
  }
});
$("download-text").addEventListener("click", () => {
  if (!result) return;
  const url = URL.createObjectURL(
    new Blob([edits.join("\n\f\n") + "\n"], {
      type: "text/plain;charset=utf-8",
    }),
  );
  const link = document.createElement("a");
  link.href = url;
  link.download = result.name.replace(/\.[^.]+$/, "") + ".txt";
  link.click();
  setTimeout(() => URL.revokeObjectURL(url), 1000);
});
window.addEventListener("beforeunload", (event) => {
  if (
    busy ||
    (result && edits.some((text, i) => text !== result.pages[i].text))
  ) {
    event.preventDefault();
    event.returnValue = "";
  }
});
(async () => {
  try {
    const response = await fetch("/api/status");
    if (!response.ok) throw Error("Server unavailable.");
    const data = await response.json();
    token = data.token;
    $("model-name").textContent = data.model;
    message("Model loaded. Everything stays on this machine.");
    controls();
  } catch {
    message(
      "Cannot reach the local OCR server. Start it with ./ocr serve, then reload this page.",
      true,
    );
  }
})();
